pages/1_orders_and_revenue.py

```diff
@@ -16,51 +16,180 @@
 ## Define data and filters. The resulting data variable includes the data with all filters applied.
 data = page_creation()
 
-st.divider()
-
-st.dataframe(data.head(10)) ## This is here just as an example. This can be deleted during development and before release.
+## Ensure all date columns are in datetime format
+date_columns = ['created_at', 'customer_created_at', 'payment_at', 'subscription_period_started_at', 'subscription_period_ended_at']
+for col in date_columns:
+    if col in data.columns:
+        data[col] = pd.to_datetime(data[col], errors='coerce')
 
 st.divider()
 
-# Everything below this point will be generated by the assigned analyst. 
-# The framework has been set, but no visualization or data processing has been applied.
-# Please perform any data processing in this file and not within the filter files. We can discuss upon completion if it makes sense to add any code to the filters file.
-
-## KPI Metrics which need to be updated.
+# KPI Metrics
 with st.container():
     col1, col2, col3, col4 = st.columns(4)
 
+    total_revenue = data['total_amount'].sum()
+    number_of_orders = data['header_id'].nunique()
+    number_of_customers = data['customer_id'].nunique()
+    min_created_at = data['created_at'].min()
+    max_created_at = data['created_at'].max()
+    new_customers = data[(data['customer_created_at'] >= min_created_at) & (data['customer_created_at'] <= max_created_at)].shape[0]
+
     with col1:
-        st.metric(label="**Total Revenue**", value="TBD")
+        st.metric(label="**Total Revenue**", value=f"${total_revenue:,.2f}")
 
     with col2:
-        st.metric(label="**Number of Orders**", value="TBD")
+        st.metric(label="**Number of Orders**", value=number_of_orders)
 
     with col3:
-        st.metric(label="**Number of Customers**", value="TBD")
+        st.metric(label="**Number of Customers in Period**", value=number_of_customers)
 
     with col4:
-        st.metric(label="**New Customers**", value="TBD")
+        st.metric(label="**New Customers**", value=new_customers)
```

Comment:
After reviewing @fivetran-catfritz PR I really liked the inclusion of the delta to these KPIs which helps tell the narrative of how the fake data is doing compared to the previous X period. 

https://github.com/fivetran/streamlit_fivetran_billing_model/blob/91011fcd634c53d16d562d7faafeb14fcec1546a/pages/2_subscriptions_report.py#L57-L61

Would you be able to add a delta to each of these metrics? I believe a Year over Year or Quarter over Quarter delta comparison would suffice here.

Reply:
Added YoYs!

Reply:
What would the LOE be to adjust this to QoQ? Just to fit into the limitation of our fake data not always having a full year to look back?

Reply:
Ah yes I had that in initially, but I was thinking it would look confusing to work with the date filter. Like if we picked June-July 2023, how does that get translated to a quarter? Same with if we did Jan-July 2023, how would a quarter get broken up? Whereas the YoY is more straightforward.

Reply:
There should be a native dataframe modification operation to group by quarter. See below as an example. 

https://github.com/fivetran/streamlit_fivetran_billing_model/blob/471ea1aceb83069a2146f1bbf3f4435660104fc2/pages/2_subscriptions_report.py#L58

Reply:
updated to include QoQ!